Description: add sunway asm support
Author: Miao Changwei <miaochangwei@uniontech.com>

---
Last-Update: 2022-05-06

--- libdrm-2.4.108.orig/xf86drm.c
+++ libdrm-2.4.108/xf86drm.c
@@ -1049,7 +1049,7 @@ static int drmOpenByBusid(const char *bu
             sv.drm_dd_major = -1;        /* Don't care */
             sv.drm_dd_minor = -1;        /* Don't care */
             if (drmSetInterfaceVersion(fd, &sv)) {
-#ifndef __alpha__
+#if !defined(__alpha__) && !defined(__sw_64__)
                 pci_domain_ok = 0;
 #endif
                 sv.drm_di_major = 1;
--- libdrm-2.4.108.orig/xf86drm.h
+++ libdrm-2.4.108/xf86drm.h
@@ -504,6 +504,34 @@ do {	register unsigned int __old __asm("
                : "cc","memory");                       \
        } while (0)
 
+#elif defined(__sw_64__)
+
+#define	DRM_CAS(lock, old, new, ret)		\
+	do {					\
+		int tmp, old32;			\
+		__asm__ __volatile__(		\
+		"	addw	$31, %5, %3\n"	\
+		"1:	lldw	%0, %2\n"	\
+		"	cmpeq	%0, %3, %1\n"	\
+		"	wr_f	%1\n"	\
+		"	beq	%1, 2f\n"	\
+		"	mov	%4, %0\n"	\
+		"	lstw	%0, %2\n"	\
+		"	rd_f	%0\n"	\
+		"	beq	%1, 2f\n"	\
+		"	beq	%0, 3f\n"	\
+		"	memb\n"			\
+		"2:	cmpeq	%1, 0, %1\n"	\
+		".subsection 2\n"		\
+		"3:	br	1b\n"		\
+		".previous"			\
+		: "=&r"(tmp), "=&r"(ret),	\
+		  "=m"(__drm_dummy_lock(lock)),	\
+		  "=&r"(old32)			\
+		: "r"(new), "r"(old)		\
+		: "memory");			\
+	} while (0)
+
 #endif /* architecture */
 #endif /* __GNUC__ >= 2 */
 
@@ -511,7 +539,7 @@ do {	register unsigned int __old __asm("
 #define DRM_CAS(lock,old,new,ret) do { ret=1; } while (0) /* FAST LOCK FAILS */
 #endif
 
-#if defined(__alpha__)
+#if defined(__alpha__) || defined(__sw_64__)
 #define DRM_CAS_RESULT(_result)		long _result
 #elif defined(__powerpc__)
 #define DRM_CAS_RESULT(_result)		int _result
